broker/src/broker.rs

```diff
@@ -0,0 +1,515 @@
+        &mut self,
+        sub_key: SubscriptionKey,
+        sub_tx: Sender<SafekeeperTimelineInfo>,
+    ) -> (SubId, Vec<Sender<SubAnnounce>>, SubAnnounce) {
+        let sub_id = self.next_sub_id;
+        self.next_sub_id += 1;
+        let announce = match sub_key {
+            SubscriptionKey::All => {
+                assert!(!self.subs_to_all.contains_key(&sub_id));
+                self.subs_to_all.insert(sub_id, sub_tx.clone());
+                SubAnnounce::AddAll(sub_tx)
+            }
+            SubscriptionKey::Timeline(ttid) => {
+                match self.subs_to_timelines.entry(ttid) {
+                    Entry::Occupied(mut o) => {
+                        let subsenders = o.get_mut();
+                        subsenders.push(SubSender(sub_id, sub_tx.clone()));
+                    }
+                    Entry::Vacant(v) => {
+                        v.insert(vec![SubSender(sub_id, sub_tx.clone())]);
+                    }
+                }
+                SubAnnounce::AddTimeline(ttid, SubSender(sub_id, sub_tx))
+            }
+        };
+        // Collect existing publishers to notify them after lock is released;
+        // TODO: the probability of channels being full here is tiny (publisher
+        // always blocks listening chan), we can try sending first and resort to
+        // cloning if needed.
+        //
+        // Deadlock is possible only if publisher tries to access shared state
+        // during its lifetime, i.e. we add maintenance of set of published
+        // tlis. Otherwise we can just await here (but lock must be replaced
+        // with Tokio one).
+        //
+        // We could also just error out if some chan is full, but that needs
+        // cleanup of incompleted job, and notifying publishers when unregistering
+        // is mandatory anyway.
+        (sub_id, self.pub_txs.values().cloned().collect(), announce)
+    }
+
+    // Unregister the subscriber. Similar to register_subscriber, returns list
+    // of channels through which publishers must be notified about the removal.
+    pub fn unregister_subscriber(
+        &mut self,
+        sub_id: SubId,
+        sub_key: SubscriptionKey,
+    ) -> Option<(Vec<Sender<SubAnnounce>>, SubAnnounce)> {
+        // We need to notify existing publishers only about per timeline
+        // subscriptions, 'all' kind is detected on its own through closed
+        // channels.
+        let announce = match sub_key {
+            SubscriptionKey::All => {
+                assert!(self.subs_to_all.contains_key(&sub_id));
+                self.subs_to_all.remove(&sub_id);
+                None
+            }
+            SubscriptionKey::Timeline(ref ttid) => {
+                remove_sub(&mut self.subs_to_timelines, ttid, sub_id);
+                Some(SubAnnounce::RemoveTimeline(*ttid, sub_id))
+            }
+        };
+        announce.map(|a| (self.pub_txs.values().cloned().collect(), a))
+    }
+
+    pub fn report(&mut self) {
```

Comment:
Should we use some logging here instead of just prints?
Also, plenty of prints in this module are commented out, that's odd.

Reply:
Sure, that's in todo list.